core-rust/state-manager/src/mempool/mempool_manager.rs

```diff
@@ -145,23 +143,36 @@ impl MempoolManager {
         max_count: usize,
         max_payload_size_bytes: u64,
     ) -> Vec<Arc<MempoolTransaction>> {
-        // TODO: don't grab whole mempool. Use a reservoir sampling method or a priority/round robin algorithm.
-        // Note that the round robin approach needs extra care to avoid relaying same transactions to same nodes.
-        let candidate_transactions = self.mempool.read().get_all_transactions();
+        let candidate_transactions = self.mempool.read().get_k_random_transactions(max_count * 2);
 
-        Self::pick_subset_with_limits(candidate_transactions, max_count, max_payload_size_bytes)
+        let mut payload_size_so_far = 0;
+        candidate_transactions
+            .into_iter()
+            .filter(|transaction| {
+                let increased_payload_size = payload_size_so_far + transaction.raw.0.len() as u64;
+                let fits = increased_payload_size <= max_payload_size_bytes;
+                if fits {
+                    payload_size_so_far = increased_payload_size;
+                }
+                fits
+            })
+            .take(max_count)
+            .collect()
     }
 
     /// Checks the committability of a random subset of transactions and removes the rejected ones
     /// from the mempool.
     /// Obeys the given limit on the number of actually executed (i.e. not cached) transactions.
     pub fn reevaluate_transaction_committability(&self, max_reevaluated_count: u32) {
-        // TODO: instead of getting whole mempool use a priority queue/round robin.
-        let mut candidate_transactions = self.mempool.read().get_all_transactions();
+        const MIN_TRANSACTIONS_TO_CHECK_FOR_REEVALUATION: u32 = 100;
+        let candidate_transactions = self.mempool.read().get_k_random_transactions(max(
```

Comment:
It feels like we should actually re-evaluate the ones which have been in here the longest since their last re-evaluation?

Reply:
Adding this to my TODO list for a next PR. :)